# REVISION DE DATA Y GENERACION DE DIM TABLES Y FACT TABLES

#### Abrimos los 6 archivos

In [1]:
import pandas as pd

In [3]:
df_SalesFinal = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/SalesFINAL12312016.csv')

In [4]:
df_PurchasesFinal = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/PurchasesFINAL12312016.csv')

In [5]:
df_BegInv = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/BegInvFINAL12312016.csv')
df_EndInv = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/EndInvFINAL12312016.csv')

In [6]:
df_PurchasePrices = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/2017PurchasePricesDec.csv')
df_InvoicePurchases = pd.read_csv('D:/OneDrive/nana/cursos/HENRY_DATA SCIENCE_DATA ANALYTICS/DATA ANALYTICS/CARRERA/MODULOS/PF_grupal/PF_final/InvoicePurchases12312016.csv')

#### Revisamos la data

In [7]:
df_PurchasesFinal.shape

(2372474, 16)

In [13]:
df_PurchasesFinal.head()

InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2015-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2015-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2016-01-02  2016-01-04  2016-02-16          35.71         6   214.26   
1    2016-01-01  2016-01-07  2016-02-21           9.35         4    37.40   
2    2016-01-02  2016-01-07  2016-02-21           9.41         5    47.05   
3    2016-01-01  2016-01-07  2016-02-21           9.35         6    56.10   
4    2016-01-02  2016-01-09  2016-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

In [14]:
df_PurchasesFinal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2372474 entries, 0 to 2372473
Data columns (total 16 columns):
 #   Column          Dtype  
---  ------          -----  
 0   InventoryId     object 
 1   Store           int64  
 2   Brand           int64  
 3   Description     object 
 4   Size            object 
 5   VendorNumber    int64  
 6   VendorName      object 
 7   PONumber        int64  
 8   PODate          object 
 9   ReceivingDate   object 
 10  InvoiceDate     object 
 11  PayDate         object 
 12  PurchasePrice   float64
 13  Quantity        int64  
 14  Dollars         float64
 15  Classification  int64  
dtypes: float64(2), int64(6), object(8)
memory usage: 289.6+ MB


In [8]:
df_SalesFinal.shape

(1048575, 14)

In [15]:
df_SalesFinal.head()

InventoryId  Store  Brand                 Description        Size  \
0  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005      1   1005     Maker's Mark Combo Pack  375mL 2 Pk   

   SalesQuantity  SalesDollars  SalesPrice SalesDate  Volume  Classification  \
0              1         16.49       16.49  1/1/2016     750               1   
1              2         32.98       16.49  1/2/2016     750               1   
2              1         16.49       16.49  1/3/2016     750               1   
3              1         14.49       14.49  1/8/2016     750               1   
4              2         69.98       34.99  1/9/2016     375               1   

   ExciseTax  VendorNo                   VendorName  
0       0.79     12546  JIM BEAM BRANDS COMPANY      
1       1.57     12546  JIM BEAM BRANDS COMPANY      
2       0.79     12546  JIM BEAM BRANDS COMPANY      
3       0.79     12546  JIM BEAM BRANDS COMPANY      
4       0.79     12546  JIM BEAM BRANDS COMPANY

In [16]:
df_SalesFinal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 14 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   InventoryId     1048575 non-null  object 
 1   Store           1048575 non-null  int64  
 2   Brand           1048575 non-null  int64  
 3   Description     1048575 non-null  object 
 4   Size            1048575 non-null  object 
 5   SalesQuantity   1048575 non-null  int64  
 6   SalesDollars    1048575 non-null  float64
 7   SalesPrice      1048575 non-null  float64
 8   SalesDate       1048575 non-null  object 
 9   Volume          1048575 non-null  int64  
 10  Classification  1048575 non-null  int64  
 11  ExciseTax       1048575 non-null  float64
 12  VendorNo        1048575 non-null  int64  
 13  VendorName      1048575 non-null  object 
dtypes: float64(3), int64(6), object(5)
memory usage: 112.0+ MB


In [9]:
df_BegInv.shape

(206529, 9)

In [17]:
df_BegInv.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2016-01-01  
1  750mL       7  10.99  2016-01-01  
2  750mL       6  36.99  2016-01-01  
3  750mL       3  38.99  2016-01-01  
4  750mL       6  34.99  2016-01-01

In [18]:
df_BegInv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206529 entries, 0 to 206528
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InventoryId  206529 non-null  object 
 1   Store        206529 non-null  int64  
 2   City         206529 non-null  object 
 3   Brand        206529 non-null  int64  
 4   Description  206529 non-null  object 
 5   Size         206529 non-null  object 
 6   onHand       206529 non-null  int64  
 7   Price        206529 non-null  float64
 8   startDate    206529 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 14.2+ MB


In [10]:
df_EndInv.shape

(224489, 9)

In [19]:
df_EndInv.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2016-12-31  
1  750mL       7  36.99  2016-12-31  
2  750mL       7  38.99  2016-12-31  
3  750mL       4  34.99  2016-12-31  
4  750mL       7  14.99  2016-12-31

In [20]:
df_EndInv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 224489 entries, 0 to 224488
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InventoryId  224489 non-null  object 
 1   Store        224489 non-null  int64  
 2   City         223205 non-null  object 
 3   Brand        224489 non-null  int64  
 4   Description  224489 non-null  object 
 5   Size         224489 non-null  object 
 6   onHand       224489 non-null  int64  
 7   Price        224489 non-null  float64
 8   endDate      224489 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 15.4+ MB


In [11]:
df_InvoicePurchases.shape

(5543, 10)

In [21]:
df_InvoicePurchases.head()

VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2016-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2016-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2016-01-09      8169   
3           480  BACARDI USA INC              2016-01-12      8106   
4           516  BANFI PRODUCTS CORP          2016-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2015-12-21  2016-02-16         6     214.26     3.47      NaN  
1  2015-12-22  2016-02-21        15     140.55     8.57      NaN  
2  2015-12-24  2016-02-16         5     106.60     4.61      NaN  
3  2015-12-20  2016-02-05     10100  137483.78  2935.20      NaN  
4  2015-12-24  2016-02-12      1935   15527.25   429.20      NaN

In [22]:
df_InvoicePurchases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5543 entries, 0 to 5542
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VendorNumber  5543 non-null   int64  
 1   VendorName    5543 non-null   object 
 2   InvoiceDate   5543 non-null   object 
 3   PONumber      5543 non-null   int64  
 4   PODate        5543 non-null   object 
 5   PayDate       5543 non-null   object 
 6   Quantity      5543 non-null   int64  
 7   Dollars       5543 non-null   float64
 8   Freight       5543 non-null   float64
 9   Approval      374 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 433.2+ KB


In [12]:
df_PurchasePrices.shape

(12261, 9)

In [23]:
df_PurchasePrices.head()

Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

In [24]:
df_PurchasePrices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Brand           12261 non-null  int64  
 1   Description     12260 non-null  object 
 2   Price           12261 non-null  float64
 3   Size            12260 non-null  object 
 4   Volume          12260 non-null  object 
 5   Classification  12261 non-null  int64  
 6   PurchasePrice   12261 non-null  float64
 7   VendorNumber    12261 non-null  int64  
 8   VendorName      12261 non-null  object 
dtypes: float64(2), int64(3), object(4)
memory usage: 862.2+ KB


#### Buscando claves primarias

In [25]:
df_BegInv['InventoryId'].value_counts() # si es clave primaria

InventoryId
79_BALLYMENA_90089    1
79_BALLYMENA_46568    1
79_BALLYMENA_46551    1
79_BALLYMENA_46489    1
79_BALLYMENA_46476    1
                     ..
1_HARDERSFIELD_75     1
1_HARDERSFIELD_72     1
1_HARDERSFIELD_63     1
1_HARDERSFIELD_62     1
1_HARDERSFIELD_60     1
Name: count, Length: 206529, dtype: int64

In [26]:
df_EndInv['InventoryId'].value_counts() # si es calve primaria

InventoryId
81_PEMBROKE_90604     1
1_HARDERSFIELD_58     1
1_HARDERSFIELD_62     1
1_HARDERSFIELD_63     1
1_HARDERSFIELD_72     1
                     ..
1_HARDERSFIELD_192    1
1_HARDERSFIELD_178    1
1_HARDERSFIELD_175    1
1_HARDERSFIELD_171    1
1_HARDERSFIELD_159    1
Name: count, Length: 224489, dtype: int64

In [28]:
df_PurchasePrices['Brand'].value_counts() # si es clave primaria

Brand
46964    1
58       1
62       1
63       1
72       1
        ..
168      1
126      1
115      1
79       1
77       1
Name: count, Length: 12261, dtype: int64

In [31]:
df_BegInv['Brand'].value_counts()

Brand
77       79
8678     79
8691     79
3344     79
3352     79
         ..
14193     1
24229     1
24161     1
23288     1
23135     1
Name: count, Length: 8094, dtype: int64

In [32]:
df_EndInv['Brand'].value_counts()

Brand
3654     80
3488     80
2122     80
3724     80
3502     80
         ..
4114      1
3648      1
1605      1
24109     1
17444     1
Name: count, Length: 9653, dtype: int64

In [29]:
df_InvoicePurchases['PONumber'].value_counts() # si es clave primaria

PONumber
13574    1
8124     1
8137     1
8169     1
8106     1
        ..
8152     1
8107     1
8151     1
8172     1
8171     1
Name: count, Length: 5543, dtype: int64

In [30]:
df_PurchasesFinal['PONumber'].value_counts()

PONumber
10936    6108
13487    5594
11300    5532
13597    5532
11191    5529
         ... 
12887       1
12928       1
12990       1
13015       1
13002       1
Name: count, Length: 5543, dtype: int64

In [33]:
df_PurchasesFinal['InvoiceDate'].value_counts()

InvoiceDate
2016-08-16    29989
2016-10-25    25231
2016-12-19    22080
2016-09-13    22057
2016-06-06    22040
              ...  
2016-03-26      438
2016-08-20      315
2016-01-04      148
2016-05-14      104
2016-08-05       60
Name: count, Length: 373, dtype: int64

In [34]:
df_InvoicePurchases['InvoiceDate'].value_counts()

InvoiceDate
2016-07-04    43
2016-07-11    35
2017-01-03    33
2016-08-08    32
2016-08-16    31
              ..
2016-08-10     4
2016-06-02     4
2016-06-25     4
2016-04-27     4
2016-11-16     4
Name: count, Length: 373, dtype: int64

In [37]:
df_PurchasesFinal['PODate'].value_counts()

PODate
2016-06-27    24752
2016-12-02    22318
2016-10-31    21681
2016-12-23    18285
2016-05-19    17458
              ...  
2015-12-22     1196
2016-10-16     1166
2015-12-25      947
2015-12-21      415
2016-10-24      346
Name: count, Length: 319, dtype: int64

In [39]:
df_InvoicePurchases['PODate'].value_counts()

PODate
2016-06-25    86
2016-09-23    76
2016-03-18    26
2016-07-01    26
2016-10-26    25
              ..
2016-03-15     9
2016-02-14     9
2016-07-06     8
2016-11-02     8
2016-08-29     7
Name: count, Length: 319, dtype: int64

In [40]:
df_InvoicePurchases['PayDate'].value_counts()

PayDate
2016-08-14    31
2016-08-12    30
2016-11-12    30
2016-08-18    29
2016-11-04    27
              ..
2017-02-18     5
2017-02-14     5
2016-02-09     3
2017-02-15     3
2016-02-08     2
Name: count, Length: 382, dtype: int64

In [41]:
df_PurchasesFinal['PayDate'].value_counts()

PayDate
2016-09-10    19713
2017-02-02    18077
2016-10-16    17693
2017-01-21    16964
2016-08-26    16310
              ...  
2016-02-08      305
2016-02-04      259
2017-02-19      111
2016-02-07      110
2017-02-15       70
Name: count, Length: 382, dtype: int64

In [42]:
df_InvoicePurchases['VendorName'].value_counts()

VendorName
AMERICAN VINTAGE BEVERAGE      55
SAZERAC CO INC                 55
BACARDI USA INC                55
BANFI PRODUCTS CORP            55
BROWN-FORMAN CORP              55
                               ..
CAPSTONE INTERNATIONAL          1
TRUETT HURST                    1
FLAVOR ESSENCE INC              1
LAUREATE IMPORTS CO             1
AAPER ALCOHOL & CHEMICAL CO     1
Name: count, Length: 129, dtype: int64

In [43]:
df_PurchasesFinal['VendorName'].value_counts()

VendorName
DIAGEO NORTH AMERICA INC       243326
JIM BEAM BRANDS COMPANY        189832
CONSTELLATION BRANDS INC       185574
MARTIGNETTI COMPANIES          176487
E & J GALLO WINERY             162567
                                ...  
CAPSTONE INTERNATIONAL              1
TRUETT HURST                        1
FLAVOR ESSENCE INC                  1
LAUREATE IMPORTS CO                 1
AAPER ALCOHOL & CHEMICAL CO         1
Name: count, Length: 129, dtype: int64

In [47]:
df_InvoicePurchases['Quantity'].value_counts()

Quantity
12       228
24       109
6         90
36        84
48        67
        ... 
6708       1
6692       1
10973      1
43240      1
8466       1
Name: count, Length: 2895, dtype: int64

In [48]:
df_PurchasesFinal['Quantity'].value_counts()

Quantity
12     627805
6      499448
11     171280
5      160343
4      129921
        ...  
750         1
602         1
912         1
422         1
552         1
Name: count, Length: 686, dtype: int64

In [49]:
df_InvoicePurchases['Dollars'].value_counts()

Dollars
214.26       22
52.08        18
120.72       12
84.30         9
428.52        8
             ..
1563.00       1
37300.48      1
202815.78     1
149007.56     1
277542.68     1
Name: count, Length: 5226, dtype: int64

In [50]:
df_PurchasesFinal['Dollars'].value_counts()

Dollars
39.96      7595
63.12      7404
85.68      6978
78.36      6320
72.84      5885
           ... 
680.74        1
132.05        1
387.03        1
722.26        1
1020.11       1
Name: count, Length: 33567, dtype: int64

In [45]:
df_InvoicePurchases['VendorNumber'].value_counts()

VendorNumber
4425      82
2000      67
1587      58
4380      55
516       55
          ..
1439       1
201359     1
9099       1
4901       1
54         1
Name: count, Length: 126, dtype: int64

In [46]:
df_PurchasesFinal['VendorNumber'].value_counts()

VendorNumber
3960      243326
12546     189832
1392      185574
4425      176781
3252      162567
           ...  
1439           1
201359         1
9099           1
4901           1
54             1
Name: count, Length: 126, dtype: int64

#### Se detecta que hay 126 VendorNumber pero 129 VendorName, se revisa y corrige en el dataframe df_InvoicePurchase

In [51]:
# Agrupar por 'VendorNumber' y contar los valores únicos de 'VendorName'
df_grouped = df_InvoicePurchases.groupby('VendorNumber')['VendorName'].nunique().reset_index()

# Filtrar los casos donde hay más de un valor único de 'VendorName'
df_multiple_vendorname = df_grouped[df_grouped['VendorName'] > 1]

# Ver cuántos VendorNumber tienen más de un VendorName
count_vendorname= df_multiple_vendorname.shape[0]

# Mostrar los VendorNumber con más de un VendorName
print(df_multiple_vendorname[['VendorNumber', 'VendorName']])

# Mostrar cuántos VendorNumber tienen más de un VendorName
print(f"Numero de VendorNumber con más de un unico VendorName: {count_vendorname}")


    VendorNumber  VendorName
19          1587           2
25          2000           2
42          4425           2
Numero de VendorNumber con más de un unico VendorName: 3


In [53]:
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_1587 = df_InvoicePurchases[df_InvoicePurchases['VendorNumber'] == 1587]

# Mostrar la fila
row_VendorNumber_1587['VendorName'].unique()


array(['VINEYARD BRANDS INC        ', 'VINEYARD BRANDS LLC        '],
      dtype=object)

In [54]:
# Filtrar la fila donde 'VendorNumber' es igual a 2000
row_VendorNumber_2000 = df_InvoicePurchases[df_InvoicePurchases['VendorNumber'] == 2000]

# Mostrar la fila
row_VendorNumber_2000['VendorName'].unique()


array(['SOUTHERN WINE & SPIRITS NE ', 'SOUTHERN GLAZERS W&S OF NE '],
      dtype=object)

In [55]:
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_4425 = df_InvoicePurchases[df_InvoicePurchases['VendorNumber'] == 4425]

# Mostrar la fila
row_VendorNumber_4425['VendorName'].unique()


array(['MARTIGNETTI COMPANIES ', 'MARTIGNETTI COMPANIES'], dtype=object)

#### Se observa que hay espacios en blanco en las tablas. Primero se corrige este error en todas las tablas

In [56]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_PurchasesFinal = df_PurchasesFinal.map(lambda x: x.strip() if isinstance(x, str) else x)

In [57]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_SalesFinal = df_SalesFinal.map(lambda x: x.strip() if isinstance(x, str) else x)

In [58]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_BegInv = df_BegInv.map(lambda x: x.strip() if isinstance(x, str) else x)

In [59]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_EndInv = df_EndInv.map(lambda x: x.strip() if isinstance(x, str) else x)

In [60]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_InvoicePurchases = df_InvoicePurchases.map(lambda x: x.strip() if isinstance(x, str) else x)

In [61]:
# Eliminar espacios en blanco al inicio y al final de cada texto en todas las columnas de texto
df_PurchasePrices = df_PurchasePrices.map(lambda x: x.strip() if isinstance(x, str) else x)

#### Revisamos nuevamente el conteo de mas de un vendor name

In [62]:
# Agrupar por 'VendorNumber' y contar los valores únicos de 'VendorName'
df_grouped = df_InvoicePurchases.groupby('VendorNumber')['VendorName'].nunique().reset_index()

# Filtrar los casos donde hay más de un valor único de 'VendorName'
df_multiple_vendorname = df_grouped[df_grouped['VendorName'] > 1]

# Ver cuántos VendorNumber tienen más de un VendorName
count_vendorname= df_multiple_vendorname.shape[0]

# Mostrar los VendorNumber con más de un VendorName
print(df_multiple_vendorname[['VendorNumber', 'VendorName']])

# Mostrar cuántos VendorNumber tienen más de un VendorName
print(f"Numero de VendorNumber con más de un unico VendorName: {count_vendorname}")


    VendorNumber  VendorName
19          1587           2
25          2000           2
Numero de VendorNumber con más de un unico VendorName: 2


#### Verificamos el nombre correcto en las tablas de inventarios y precios de compra

In [63]:
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_1587 = df_PurchasePrices[df_PurchasePrices['VendorNumber'] == 1587]

# Mostrar la fila
row_VendorNumber_1587['VendorName'].unique()


array(['VINEYARD BRANDS INC', 'VINEYARD BRANDS LLC'], dtype=object)

In [66]:
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_1587 = df_SalesFinal[df_SalesFinal['VendorNo'] == 1587]

# Mostrar la fila
row_VendorNumber_1587['VendorName'].unique()


array(['VINEYARD BRANDS INC'], dtype=object)

In [68]:
# Filtrar la fila donde 'VendorName' es igual a 'VINEYARD BRANDS INC'
row_VendorName_1 = df_InvoicePurchases[df_InvoicePurchases['VendorName'] == 'VINEYARD BRANDS INC']

# Verificar la cantidad de compras
row_VendorName_1['InvoiceDate'].unique()


array(['2016-01-04', '2016-01-11', '2016-01-18', '2016-02-02',
       '2016-02-03', '2016-02-09', '2016-02-22', '2016-03-01',
       '2016-03-08', '2016-03-11', '2016-03-21', '2016-03-28',
       '2016-04-01', '2016-04-08', '2016-04-15', '2016-04-18',
       '2016-04-30', '2016-05-07', '2016-05-12', '2016-05-23',
       '2016-05-28', '2016-05-31', '2016-06-13', '2016-06-19',
       '2016-06-27', '2016-07-01', '2016-07-04', '2016-07-07',
       '2016-07-12', '2016-07-23', '2016-07-30', '2016-08-08',
       '2016-08-12', '2016-08-16', '2016-08-30', '2016-08-31',
       '2016-09-06', '2016-09-20', '2016-09-23', '2016-10-01',
       '2016-10-08', '2016-10-07', '2016-10-10', '2016-10-23',
       '2016-10-27', '2016-11-04', '2016-11-12', '2016-11-15',
       '2016-11-29', '2016-12-06', '2016-12-09', '2016-12-18',
       '2016-12-25', '2016-12-29', '2017-01-03'], dtype=object)

In [67]:
# Filtrar la fila donde 'VendorName' es igual a 'VINEYARD BRANDS LLC'
row_VendorName_1 = df_InvoicePurchases[df_InvoicePurchases['VendorName'] == 'VINEYARD BRANDS LLC']

# Verificar la cantidad de compras
row_VendorName_1['InvoiceDate'].unique()


array(['2016-12-05', '2016-12-25', '2017-01-03'], dtype=object)

#### Corregimos el nombre en df_InvoicePurchase en vendor que ha tenido menos compras (nombre erróneo)

In [69]:
# Verificar si el nombre incorrecto está en la columna VendorName y corregirlo
df_InvoicePurchases.loc[df_InvoicePurchases['VendorName'] == 'VINEYARD BRANDS LLC', 'VendorName'] = 'VINEYARD BRANDS INC'


In [70]:
# Revisamos nuevamente
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_1587 = df_InvoicePurchases[df_InvoicePurchases['VendorNumber'] == 1587]

# Mostrar la fila
row_VendorNumber_1587['VendorName'].unique()


array(['VINEYARD BRANDS INC'], dtype=object)

#### Repetimos el paso para el otro nombre

In [73]:
# Filtrar la fila donde 'VendorNumber' es igual a 2000
row_VendorNumber_2000 = df_PurchasePrices[df_PurchasePrices['VendorNumber'] == 2000]

# Mostrar la fila
row_VendorNumber_2000['VendorName'].unique()


array(['SOUTHERN WINE & SPIRITS NE', 'SOUTHERN GLAZERS W&S OF NE'],
      dtype=object)

In [75]:
# Filtrar la fila donde 'VendorNumber' es igual a 2000
row_VendorNumber_2000 = df_SalesFinal[df_SalesFinal['VendorNo'] == 2000]

# Mostrar la fila
row_VendorNumber_2000['VendorName'].unique()


array(['SOUTHERN WINE & SPIRITS NE'], dtype=object)

In [72]:
# Filtrar la fila donde 'VendorName' es igual a 'SOUTHERN WINE & SPIRITS NE'
row_VendorName_2000 = df_InvoicePurchases[df_InvoicePurchases['VendorName'] == 'SOUTHERN WINE & SPIRITS NE']

# Verificar la cantidad de compras
row_VendorName_2000['InvoiceDate'].unique()


array(['2016-01-05', '2016-01-17', '2016-01-21', '2016-01-31',
       '2016-02-02', '2016-02-10', '2016-02-19', '2016-03-01',
       '2016-02-29', '2016-03-07', '2016-03-15', '2016-03-28',
       '2016-04-01', '2016-04-05', '2016-04-16', '2016-04-19',
       '2016-04-28', '2016-05-08', '2016-05-09', '2016-05-18',
       '2016-05-25', '2016-06-06', '2016-06-07', '2016-06-19',
       '2016-06-23', '2016-07-05', '2016-07-04', '2016-07-08',
       '2016-07-12', '2016-07-20', '2016-07-26', '2016-08-08',
       '2016-08-11', '2016-08-16', '2016-08-22', '2016-08-31',
       '2016-09-05', '2016-09-13', '2016-09-26', '2016-09-29',
       '2016-10-08', '2016-10-07', '2016-10-11', '2016-10-22',
       '2016-10-29', '2016-11-02', '2016-11-12', '2016-11-19',
       '2016-11-28', '2016-12-01', '2016-12-08', '2016-12-17',
       '2016-12-26', '2016-12-30', '2017-01-08'], dtype=object)

In [76]:
# Filtrar la fila donde 'VendorName' es igual a 'SOUTHERN GLAZERS W&S OF NE'
row_VendorName_2000 = df_InvoicePurchases[df_InvoicePurchases['VendorName'] == 'SOUTHERN GLAZERS W&S OF NE']

# Verificar la cantidad de compras
row_VendorName_2000['InvoiceDate'].unique()


array(['2016-09-17', '2016-09-24', '2016-09-27', '2016-10-07',
       '2016-10-16', '2016-10-23', '2016-11-01', '2016-11-07',
       '2016-11-26', '2016-11-30', '2016-12-12', '2016-12-27'],
      dtype=object)

#### Corregimos el nombre en df_InvoicePurchase en vendor que ha tenido menos compras (nombre erróneo)

In [77]:
# Verificar si el nombre incorrecto está en la columna VendorName y corregirlo
df_InvoicePurchases.loc[df_InvoicePurchases['VendorName'] == 'SOUTHERN GLAZERS W&S OF NE', 'VendorName'] = 'SOUTHERN WINE & SPIRITS NE'


In [78]:
# Filtrar la fila donde 'VendorNumber' es igual a 2000
row_VendorNumber_2000 = df_InvoicePurchases[df_InvoicePurchases['VendorNumber'] == 2000]

# Mostrar la fila
row_VendorNumber_2000['VendorName'].unique()


array(['SOUTHERN WINE & SPIRITS NE'], dtype=object)

### Analizamos las tablas para definir la normalizacion

In [80]:
df_InvoicePurchases.shape

(5543, 10)

In [81]:
df_InvoicePurchases['Freight'].value_counts()

Freight
0.26       13
0.23       11
0.60       11
0.52       10
0.44        9
           ..
2027.87     1
127.05      1
408.72      1
5972.45     1
321.02      1
Name: count, Length: 4052, dtype: int64

In [79]:
df_InvoicePurchases['Dollars'].value_counts()

Dollars
214.26       22
52.08        18
120.72       12
84.30         9
428.52        8
             ..
1563.00       1
37300.48      1
202815.78     1
149007.56     1
277542.68     1
Name: count, Length: 5226, dtype: int64

In [83]:
df_InvoicePurchases['InvoiceDate'].value_counts()

InvoiceDate
2016-07-04    43
2016-07-11    35
2017-01-03    33
2016-08-08    32
2016-08-16    31
              ..
2016-08-10     4
2016-06-02     4
2016-06-25     4
2016-04-27     4
2016-11-16     4
Name: count, Length: 373, dtype: int64

In [84]:
df_InvoicePurchases['PODate'].value_counts()

PODate
2016-06-25    86
2016-09-23    76
2016-03-18    26
2016-07-01    26
2016-10-26    25
              ..
2016-03-15     9
2016-02-14     9
2016-07-06     8
2016-11-02     8
2016-08-29     7
Name: count, Length: 319, dtype: int64

In [85]:
df_InvoicePurchases['PONumber'].value_counts()

PONumber
13574    1
8124     1
8137     1
8169     1
8106     1
        ..
8152     1
8107     1
8151     1
8172     1
8171     1
Name: count, Length: 5543, dtype: int64

In [87]:
df_InvoicePurchases['PayDate'].value_counts()

PayDate
2016-08-14    31
2016-08-12    30
2016-11-12    30
2016-08-18    29
2016-11-04    27
              ..
2017-02-18     5
2017-02-14     5
2016-02-09     3
2017-02-15     3
2016-02-08     2
Name: count, Length: 382, dtype: int64

In [88]:
df_InvoicePurchases['Quantity'].value_counts()

Quantity
12       228
24       109
6         90
36        84
48        67
        ... 
6708       1
6692       1
10973      1
43240      1
8466       1
Name: count, Length: 2895, dtype: int64

In [89]:
df_InvoicePurchases['VendorName'].value_counts()

VendorName
MARTIGNETTI COMPANIES          82
SOUTHERN WINE & SPIRITS NE     67
VINEYARD BRANDS INC            58
CHARLES JACQUIN ET CIE INC     55
BANFI PRODUCTS CORP            55
                               ..
CAPSTONE INTERNATIONAL          1
FLAVOR ESSENCE INC              1
TRUETT HURST                    1
LAUREATE IMPORTS CO             1
AAPER ALCOHOL & CHEMICAL CO     1
Name: count, Length: 126, dtype: int64

In [90]:
df_InvoicePurchases['VendorNumber'].value_counts()

VendorNumber
4425      82
2000      67
1587      58
4380      55
516       55
          ..
1439       1
201359     1
9099       1
4901       1
54         1
Name: count, Length: 126, dtype: int64

### Guardamos nuestros archivos con las correcciones hechas hasta este momento

In [91]:
df_InvoicePurchases.to_csv('InvoicePurchases.csv', index=False)
df_PurchasePrices.to_csv('PurchasePrices.csv', index=False)
df_BegInv.to_csv('BegInv.csv', index=False)
df_EndInv.to_csv('EndInv.csv', index=False)
df_SalesFinal.to_csv('SalesFinal.csv', index=False)
df_PurchasesFinal.to_csv('PurchasesFinal.csv', index=False)


#### Volvemos a abrir

In [92]:
import pandas as pd

In [93]:
df_InvoicePurchases= pd.read_csv('InvoicePurchases.csv')
df_PurchasePrices =pd.read_csv('PurchasePrices.csv')
df_BegInv= pd.read_csv('BegInv.csv')
df_EndInv= pd.read_csv('EndInv.csv')
df_SalesFinal= pd.read_csv('SalesFinal.csv')
df_PurchasesFinal = pd.read_csv('PurchasesFinal.csv')


In [94]:
df_InvoicePurchases['Approval'].value_counts()

Approval
Frank Delahunt    374
Name: count, dtype: int64

In [99]:
df_InvoicePurchases['Approval'].unique()

array([nan, 'Frank Delahunt'], dtype=object)

In [95]:
# Agrupar por 'InvoiceDate' y contar los valores únicos de 'Approval'
df_grouped = df_InvoicePurchases.groupby('InvoiceDate')['Approval'].nunique().reset_index()

# Filtrar los casos donde hay más de un valor único de 'Approval'
df_multiple_approval = df_grouped[df_grouped['Approval'] > 1]

# Ver cuántos InvoiceDate tienen más de un Approval
count_approval= df_multiple_approval.shape[0]

# Mostrar los InvoiceDate con más de un Approval
print(df_multiple_approval[['InvoiceDate', 'Approval']])

# Mostrar cuántos InvoiceDate tienen más de un Approval
print(f"Numero de InvoiceDate con más de un unico Approval: {count_approval}")


Empty DataFrame
Columns: [InvoiceDate, Approval]
Index: []
Numero de InvoiceDate con más de un unico Approval: 0


In [100]:
df_SalesFinal['VendorNo'].value_counts()

VendorNo
3960      125869
12546      94057
4425       76625
1392       70792
3252       65240
           ...  
172662         1
9710           1
1265           1
2              1
1002           1
Name: count, Length: 116, dtype: int64

In [101]:
df_PurchasesFinal['VendorNumber'].value_counts()

VendorNumber
3960      243326
12546     189832
1392      185574
4425      176781
3252      162567
           ...  
1439           1
201359         1
9099           1
4901           1
54             1
Name: count, Length: 126, dtype: int64

In [102]:
df_InvoicePurchases['VendorNumber'].value_counts()

VendorNumber
4425      82
2000      67
1587      58
4380      55
516       55
          ..
1439       1
201359     1
9099       1
4901       1
54         1
Name: count, Length: 126, dtype: int64

### Creamos la Tabla DimVendor de la Tabla InvoicePurchase

In [105]:
df_InvoicePurchases.columns

Index(['VendorNumber', 'VendorName', 'InvoiceDate', 'PONumber', 'PODate',
       'PayDate', 'Quantity', 'Dollars', 'Freight', 'Approval'],
      dtype='object')

In [146]:
# Extraer valores únicos de algunas de las columnas de InvoicePrice
DimVendor = df_InvoicePurchases[['VendorNumber', 'VendorName']].drop_duplicates().reset_index(drop=True)


In [148]:
DimVendor.shape

(126, 2)

In [147]:
DimVendor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   VendorNumber  126 non-null    int64 
 1   VendorName    126 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.1+ KB


In [149]:
DimVendor['VendorNumber'].value_counts()

VendorNumber
105       1
4466      1
388       1
480       1
516       1
         ..
201359    1
4901      1
90059     1
5083      1
54        1
Name: count, Length: 126, dtype: int64

In [150]:
DimVendor['VendorName'].value_counts()

VendorName
ALTAMAR BRANDS LLC             1
AMERICAN VINTAGE BEVERAGE      1
ATLANTIC IMPORTING COMPANY     1
BACARDI USA INC                1
BANFI PRODUCTS CORP            1
                              ..
FLAVOR ESSENCE INC             1
LAUREATE IMPORTS CO            1
BLACK COVE BEVERAGES           1
LOYAL DOG WINERY               1
AAPER ALCOHOL & CHEMICAL CO    1
Name: count, Length: 126, dtype: int64

#### Reviso que no tenga datos nulos

In [151]:
total_nulos_vendor = DimVendor.isnull().sum()[DimVendor.isnull().sum()>0]
print(total_nulos_vendor)

Series([], dtype: int64)


### Creamos el ID para DimVendor

In [152]:
# Generar VendroID en DimVendor
DimVendor['VendorID'] = DimVendor.index + 1

In [153]:
DimVendor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   VendorNumber  126 non-null    int64 
 1   VendorName    126 non-null    object
 2   VendorID      126 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 3.1+ KB


In [154]:
DimVendor.dtypes

VendorNumber     int64
VendorName      object
VendorID         int64
dtype: object

### Ahora revisamos la Tabla InvoicePurchase y agregamos el VendorID

#### Primero Eliminamos la columna approval que no es util para nuestros cálculos

In [155]:
# Elimino la columna 'Approval'
df_InvoicePurchases=df_InvoicePurchases.drop(columns=['Approval'])

In [156]:
DimVendor.columns

Index(['VendorNumber', 'VendorName', 'VendorID'], dtype='object')

In [157]:
df_InvoicePurchases.columns

Index(['VendorNumber', 'VendorName', 'InvoiceDate', 'PONumber', 'PODate',
       'PayDate', 'Quantity', 'Dollars', 'Freight'],
      dtype='object')

In [158]:
df_InvoicePurchases.shape

(5543, 9)

In [171]:
df_InvoicePurchases['InvoiceDate'].value_counts()

InvoiceDate
2016-07-04    43
2016-07-11    35
2017-01-03    33
2016-08-08    32
2016-08-16    31
              ..
2016-08-10     4
2016-06-02     4
2016-06-25     4
2016-04-27     4
2016-11-16     4
Name: count, Length: 373, dtype: int64

In [169]:
# Contar los valores únicos en la columna 'InvoiceDate'
unique_dates = df_InvoicePurchases['InvoiceDate'].nunique()

# Mostrar el resultado
print(f'La columna "InvoiceDate" tiene {unique_dates} valores únicos.')


La columna "InvoiceDate" tiene 373 valores únicos.


#### Insertamos VendorID a la tabla InvoicePurchase y le cambiamos el nombre por Ordenes_Compra

In [159]:
# Realizar un merge entre InvoicePurchase con DimVendor basándose en la columna VendorNumber
Ordenes_Compra = df_InvoicePurchases.merge(DimVendor[['VendorID','VendorNumber']], left_on='VendorNumber', right_on='VendorNumber', how='left')


In [160]:
Ordenes_Compra.columns

Index(['VendorNumber', 'VendorName', 'InvoiceDate', 'PONumber', 'PODate',
       'PayDate', 'Quantity', 'Dollars', 'Freight', 'VendorID'],
      dtype='object')

### Eliminamos columnas redundantes

In [161]:
# Elimino la columna 'VendorNumber' y 'VendorName'
Ordenes_Compra = Ordenes_Compra.drop(columns=['VendorNumber','VendorName'])


In [162]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'VendorID'],
      dtype='object')

In [164]:
Ordenes_Compra.shape

(5543, 8)

In [165]:
Ordenes_Compra['PONumber'].value_counts()


PONumber
13574    1
8124     1
8137     1
8169     1
8106     1
        ..
8152     1
8107     1
8151     1
8172     1
8171     1
Name: count, Length: 5543, dtype: int64

In [167]:
Ordenes_Compra['PONumber'].unique

<bound method Series.unique of 0        8124
1        8137
2        8169
3        8106
4        8170
        ...  
5538    13626
5539    13661
5540    13643
5541    13602
5542    13574
Name: PONumber, Length: 5543, dtype: int64>

#### Reviso nulos

In [163]:
total_nulos_oc = Ordenes_Compra.isnull().sum()[Ordenes_Compra.isnull().sum()>0]
print(total_nulos_oc)

Series([], dtype: int64)


In [ ]:
# Elimino la columna 'VendorNumber' y 'VendorName'
Ordenes_Compra = Ordenes_Compra.drop(columns=['VendorNumber','VendorName'])


### Ahora revisamos la Tabla PurchasesFinal y agregamos el OC_ID y le cambiamos el nombre a Fact_Detalles_Compra

#### Creamos el OrdenCompraID en la nueva dataframe Ordenes_Compra

In [176]:
# Generar OrdenCompraID en Ordenes_Compra
Ordenes_Compra['OrdenCompraID'] = Ordenes_Compra.index + 1

In [173]:
df_PurchasesFinal['PONumber'].value_counts()

PONumber
10936    6108
13487    5594
11300    5532
13597    5532
11191    5529
         ... 
12887       1
12928       1
12990       1
13015       1
13002       1
Name: count, Length: 5543, dtype: int64

In [175]:
Ordenes_Compra['PONumber'].value_counts()

PONumber
13574    1
8124     1
8137     1
8169     1
8106     1
        ..
8152     1
8107     1
8151     1
8172     1
8171     1
Name: count, Length: 5543, dtype: int64

In [177]:
df_PurchasesFinal.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification'],
      dtype='object')

In [178]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'VendorID', 'OrdenCompraID'],
      dtype='object')

In [179]:
# Realizar un merge entre PurchaseFinal y Ordenes_Compra basándose en la columna PONumber
# y cambiamos el nombre de la tabla PurchasesFinal
Fact_Detalle_Compras = df_PurchasesFinal.merge(Ordenes_Compra[['PONumber', 'OrdenCompraID']], left_on='PONumber', right_on='PONumber', how='left')


In [180]:
Fact_Detalle_Compras['OrdenCompraID'].value_counts()

OrdenCompraID
2823    6108
5356    5594
3130    5532
5498    5532
3026    5529
        ... 
4802       1
4810       1
4908       1
4835       1
4859       1
Name: count, Length: 5543, dtype: int64

In [182]:
Fact_Detalle_Compras['PONumber'].value_counts()

PONumber
10936    6108
13487    5594
11300    5532
13597    5532
11191    5529
         ... 
12887       1
12928       1
12990       1
13015       1
13002       1
Name: count, Length: 5543, dtype: int64

In [181]:
Fact_Detalle_Compras.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification',
       'OrdenCompraID'],
      dtype='object')

In [183]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'VendorID', 'OrdenCompraID'],
      dtype='object')

### Borramos columnas redundantes

In [184]:
# Elimino las columnas redundantes
Fact_Detalle_Compras = Fact_Detalle_Compras.drop(columns=['VendorNumber', 'VendorName', 'PONumber', 'PODate', 'InvoiceDate','PayDate'])


In [185]:
Fact_Detalle_Compras.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'ReceivingDate',
       'PurchasePrice', 'Quantity', 'Dollars', 'Classification',
       'OrdenCompraID'],
      dtype='object')

In [186]:
Fact_Detalle_Compras.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2372474 entries, 0 to 2372473
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   InventoryId     object 
 1   Store           int64  
 2   Brand           int64  
 3   Description     object 
 4   Size            object 
 5   ReceivingDate   object 
 6   PurchasePrice   float64
 7   Quantity        int64  
 8   Dollars         float64
 9   Classification  int64  
 10  OrdenCompraID   int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 199.1+ MB


#### Revisamos nulos

In [187]:
total_nulos_detalle_oc = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos_detalle_oc)

Size    3
dtype: int64


In [188]:
print(Fact_Detalle_Compras[Fact_Detalle_Compras['Size'].isnull()])

                 InventoryId  Store  Brand                 Description Size  \
1109668    34_PITMERDEN_3121     34   3121    Pinnacle Rainbow Sherbet  NaN   
1112426    34_PITMERDEN_5678     34   5678      Skinnygirl Pina Colada  NaN   
1116302  39_EASTHALLOW_15365     39  15365  Alabaster 07 Tinta de Toro  NaN   

        ReceivingDate  PurchasePrice  Quantity  Dollars  Classification  \
1109668    2016-07-04           6.93         7    48.51               1   
1112426    2016-07-09           6.93         6    41.58               1   
1116302    2016-07-07          91.83         1    91.83               2   

         OrdenCompraID  
1109668           2839  
1112426           2839  
1116302           2846  


### Guardamos nuestras nuevas tablas DimVendor, Ordenes_Compra y Fact_Detalle_Compras

In [189]:
Ordenes_Compra.to_csv('Dim_OrdenesCompra.csv', index=False)
DimVendor.to_csv('Dim_Vendor.csv', index=False)
Fact_Detalle_Compras.to_csv('Fact_DetalleCompras.csv', index=False)


#### Abrimos nuestros nuevos archivos

In [191]:
Ordenes_Compra = pd.read_csv('Dim_OrdenesCompra.csv')
DimVendor = pd.read_csv('Dim_Vendor.csv')
Fact_Detalle_Compras = pd.read_csv('Fact_DetalleCompras.csv')

### Seguimos con la limpieza de Fact_Detalle_Compras

In [192]:
total_nulos_detalle_compras = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos_detalle_compras)

Size    3
dtype: int64


In [193]:
Fact_Detalle_Compras[Fact_Detalle_Compras['Size'].isnull()]

InventoryId  Store  Brand                 Description Size  \
1109668    34_PITMERDEN_3121     34   3121    Pinnacle Rainbow Sherbet  NaN   
1112426    34_PITMERDEN_5678     34   5678      Skinnygirl Pina Colada  NaN   
1116302  39_EASTHALLOW_15365     39  15365  Alabaster 07 Tinta de Toro  NaN   

        ReceivingDate  PurchasePrice  Quantity  Dollars  Classification  \
1109668    2016-07-04           6.93         7    48.51               1   
1112426    2016-07-09           6.93         6    41.58               1   
1116302    2016-07-07          91.83         1    91.83               2   

         OrdenCompraID  
1109668           2839  
1112426           2839  
1116302           2846

### Iniciamos la inspección de PurchasePrices

In [194]:
df_PurchasePrices.shape

(12261, 9)

In [195]:
df_PurchasePrices['Brand'].value_counts() # es la clave primaria

Brand
46964    1
58       1
62       1
63       1
72       1
        ..
168      1
126      1
115      1
79       1
77       1
Name: count, Length: 12261, dtype: int64

In [196]:
df_PurchasePrices.head()

Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                  VendorName  
0           9.28          8320     SHAW ROSS INT L IMP LTD  
1          28.67          1128           BROWN-FORMAN CORP  
2          30.46          1128           BROWN-FORMAN CORP  
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP  
4          10.94          7245        PROXIMO SPIRITS INC.

In [197]:
df_PurchasePrices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Brand           12261 non-null  int64  
 1   Description     12260 non-null  object 
 2   Price           12261 non-null  float64
 3   Size            12260 non-null  object 
 4   Volume          12260 non-null  object 
 5   Classification  12261 non-null  int64  
 6   PurchasePrice   12261 non-null  float64
 7   VendorNumber    12261 non-null  int64  
 8   VendorName      12261 non-null  object 
dtypes: float64(2), int64(3), object(4)
memory usage: 862.2+ KB


In [198]:
total_nulos_pp = df_PurchasePrices.isnull().sum()[df_PurchasePrices.isnull().sum()>0]
print(total_nulos_pp)

Description    1
Size           1
Volume         1
dtype: int64


#### Se identifica la fila con valores vacios

In [199]:
df_PurchasePrices[df_PurchasePrices['Size'].isnull()]

Brand Description  Price Size Volume  Classification  PurchasePrice  \
7915   4202         NaN    0.0  NaN    NaN               1          11.19   

      VendorNumber       VendorName  
7915           480  BACARDI USA INC

#### Reviso los datos correctos en las tablas de Inventario y ventas y compras que tambien tienen Brand

In [200]:
# Filtrar la fila donde 'Brand' es igual a 4202
row_brand_4202 = df_SalesFinal[df_SalesFinal['Brand'] == 4202]

# Mostrar la fila
print(row_brand_4202)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, SalesQuantity, SalesDollars, SalesPrice, SalesDate, Volume, Classification, ExciseTax, VendorNo, VendorName]
Index: []


In [201]:
# Filtrar la fila donde 'Brand' es igual a 4202
row_brand_4202 = df_BegInv[df_BegInv['Brand'] == 4202]

# Mostrar la fila
print(row_brand_4202)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, startDate]
Index: []


In [202]:
# Filtrar la fila donde 'Brand' es igual a 4202
row_brand_4202 = df_EndInv[df_EndInv['Brand'] == 4202]

# Mostrar la fila
print(row_brand_4202)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, endDate]
Index: []


In [203]:
# Filtrar la fila donde 'Brand' es igual a 4202
row_brand_4202 = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'] == 4202]

# Mostrar la fila
print(row_brand_4202)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, ReceivingDate, PurchasePrice, Quantity, Dollars, Classification, OrdenCompraID]
Index: []


#### Conclusion: no se encuentran datos para Size = 4202

#### Se elimina la fila que contiene Brand = 4202

In [204]:
# Eliminar filas donde 'Brand' es igual a 4202
df_PurchasePrices = df_PurchasePrices[df_PurchasePrices['Brand'] != 4202]

# Mostrar el DataFrame actualizado (opcional)
df_PurchasePrices[df_PurchasePrices['Brand'].isnull()]


Empty DataFrame
Columns: [Brand, Description, Price, Size, Volume, Classification, PurchasePrice, VendorNumber, VendorName]
Index: []

In [205]:
total_nulos_pp = df_PurchasePrices.isnull().sum()[df_PurchasePrices.isnull().sum()>0]
print(total_nulos_pp)

Series([], dtype: int64)


In [207]:
total_nulos_pp = df_PurchasePrices.isnull().sum()[df_PurchasePrices.isnull().sum()>0]
print(total_nulos_pp)

Series([], dtype: int64)


In [210]:
# Detectar valores NaN en la columna 'Size'
nan_rows = df_PurchasePrices[df_PurchasePrices['Size'].isna()]

# Mostrar las filas que contienen NaN en la columna 'Size'
print(nan_rows)


Empty DataFrame
Columns: [Brand, Description, Price, Size, Volume, Classification, PurchasePrice, VendorNumber, VendorName]
Index: []


In [212]:
# Detectar valores no numéricos en la columna 'Size'
non_numeric_rows = df_PurchasePrices[pd.to_numeric(df_PurchasePrices['Volume'], errors='coerce').isna()]

# Mostrar las filas con valores no numéricos en la columna 'Size'
print(non_numeric_rows)


       Brand                   Description  Price     Size   Volume  \
542     2993             Angostura Bitters   7.49  Unknown  Unknown   
5921    9908      Tito's Copper Mug 2 Pack  21.01  Unknown  Unknown   
8795    8992                      Group 92   1.99  Unknown  Unknown   
10009  90590  Overture Champagne 2Glass Pk  19.95  Unknown  Unknown   

       Classification  PurchasePrice  VendorNumber             VendorName  
542                 1           5.39          5895  Mizkan Americas, Inc.  
5921                1          16.15          4425  MARTIGNETTI COMPANIES  
8795                1           1.43          1703   ALISA CARR BEVERAGES  
10009               2          13.12          4425  MARTIGNETTI COMPANIES  


#### Busco en otras tablas los valores correctos de Size y Volumen para Brand 2993, 9908, 8992 y 90590

In [213]:
# Filtrar la fila donde 'Brand' es igual a 2993
row_brand_2993 = df_EndInv[df_EndInv['Brand'] == 2993]

# Mostrar la fila
print(row_brand_2993)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, endDate]
Index: []


In [214]:
# Filtrar la fila donde 'Brand' es igual a 2993
row_brand_2993 = df_BegInv[df_BegInv['Brand'] == 2993]

# Mostrar la fila
print(row_brand_2993)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, startDate]
Index: []


In [215]:
# Filtrar la fila donde 'Brand' es igual a 2993
row_brand_2993 = df_SalesFinal[df_SalesFinal['Brand'] == 2993]

# Mostrar la fila
print(row_brand_2993)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, SalesQuantity, SalesDollars, SalesPrice, SalesDate, Volume, Classification, ExciseTax, VendorNo, VendorName]
Index: []


In [216]:
# Filtrar la fila donde 'Brand' es igual a 2993
row_brand_2993 = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'] == 2993]

# Mostrar la fila
print(row_brand_2993)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, ReceivingDate, PurchasePrice, Quantity, Dollars, Classification, OrdenCompraID]
Index: []


#### No se encontró informacion sobre Brand 2993

In [217]:
# Filtrar la fila donde 'Brand' es igual a 9908
row_brand_9908 = df_EndInv[df_EndInv['Brand'] == 9908]

# Mostrar la fila
print(row_brand_9908)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, endDate]
Index: []


In [218]:
# Filtrar la fila donde 'Brand' es igual a 9908
row_brand_9908 = df_BegInv[df_BegInv['Brand'] == 9908]

# Mostrar la fila
print(row_brand_9908)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, startDate]
Index: []


In [219]:
# Filtrar la fila donde 'Brand' es igual a 9908
row_brand_9908 = df_SalesFinal[df_SalesFinal['Brand'] == 9908]

# Mostrar la fila
print(row_brand_9908)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, SalesQuantity, SalesDollars, SalesPrice, SalesDate, Volume, Classification, ExciseTax, VendorNo, VendorName]
Index: []


In [220]:
# Filtrar la fila donde 'Brand' es igual a 9908
row_brand_9908 = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'] == 9908]

# Mostrar la fila
print(row_brand_9908)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, ReceivingDate, PurchasePrice, Quantity, Dollars, Classification, OrdenCompraID]
Index: []


#### No se encontró informacion sobre Brand 9908

In [221]:
# Filtrar la fila donde 'Brand' es igual a 8992
row_brand_8992 = df_EndInv[df_EndInv['Brand'] == 8992]

# Mostrar la fila
print(row_brand_8992)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, endDate]
Index: []


In [222]:
# Filtrar la fila donde 'Brand' es igual a 8992
row_brand_8992 = df_BegInv[df_BegInv['Brand'] == 8992]

# Mostrar la fila
print(row_brand_8992)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, startDate]
Index: []


In [223]:
# Filtrar la fila donde 'Brand' es igual a 8992
row_brand_8992 = df_SalesFinal[df_SalesFinal['Brand'] == 8992]

# Mostrar la fila
print(row_brand_8992)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, SalesQuantity, SalesDollars, SalesPrice, SalesDate, Volume, Classification, ExciseTax, VendorNo, VendorName]
Index: []


In [224]:
# Filtrar la fila donde 'Brand' es igual a 8992
row_brand_8992 = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'] == 8992]

# Mostrar la fila
print(row_brand_8992)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, ReceivingDate, PurchasePrice, Quantity, Dollars, Classification, OrdenCompraID]
Index: []


#### No se encontró informacion sobre Brand 8992

In [225]:
# Filtrar la fila donde 'Brand' es igual a 90590
row_brand_90590 = df_EndInv[df_EndInv['Brand'] == 90590]

# Mostrar la fila
print(row_brand_90590)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, endDate]
Index: []


In [226]:
# Filtrar la fila donde 'Brand' es igual a 90590
row_brand_90590 = df_BegInv[df_BegInv['Brand'] == 90590]

# Mostrar la fila
print(row_brand_90590)


Empty DataFrame
Columns: [InventoryId, Store, City, Brand, Description, Size, onHand, Price, startDate]
Index: []


In [227]:
# Filtrar la fila donde 'Brand' es igual a 90590
row_brand_90590 = df_SalesFinal[df_SalesFinal['Brand'] == 90590]

# Mostrar la fila
print(row_brand_90590)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, SalesQuantity, SalesDollars, SalesPrice, SalesDate, Volume, Classification, ExciseTax, VendorNo, VendorName]
Index: []


In [228]:
# Filtrar la fila donde 'Brand' es igual a 90590
row_brand_90590 = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'] == 90590]

# Mostrar la fila
print(row_brand_90590)


Empty DataFrame
Columns: [InventoryId, Store, Brand, Description, Size, ReceivingDate, PurchasePrice, Quantity, Dollars, Classification, OrdenCompraID]
Index: []


#### No se encontró informacion sobre Brand 90590

#### Conclusion: borramos todas las filas con esas brand inexistentes

In [229]:
# Eliminar las filas con los valores 2993, 9908, 8992 y 90590 en la columna 'Brand'
df_PurchasePrices = df_PurchasePrices[~df_PurchasePrices['Brand'].isin([2993, 9908, 8992, 90590])]


In [230]:
# Mostrar el DataFrame actualizado
df_PurchasePrices[df_PurchasePrices['Brand'].isnull()]

Empty DataFrame
Columns: [Brand, Description, Price, Size, Volume, Classification, PurchasePrice, VendorNumber, VendorName]
Index: []

In [231]:
df_PurchasePrices.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12256 entries, 0 to 12260
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Brand           12256 non-null  int64  
 1   Description     12256 non-null  object 
 2   Price           12256 non-null  float64
 3   Size            12256 non-null  object 
 4   Volume          12256 non-null  object 
 5   Classification  12256 non-null  int64  
 6   PurchasePrice   12256 non-null  float64
 7   VendorNumber    12256 non-null  int64  
 8   VendorName      12256 non-null  object 
dtypes: float64(2), int64(3), object(4)
memory usage: 957.5+ KB


#### Regresamos a analizar la tabla Detalle_Compras

In [232]:
total_nulos_detalle_oc = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos_detalle_oc)


Size    3
dtype: int64


In [233]:
Fact_Detalle_Compras[Fact_Detalle_Compras['Size'].isnull()]

InventoryId  Store  Brand                 Description Size  \
1109668    34_PITMERDEN_3121     34   3121    Pinnacle Rainbow Sherbet  NaN   
1112426    34_PITMERDEN_5678     34   5678      Skinnygirl Pina Colada  NaN   
1116302  39_EASTHALLOW_15365     39  15365  Alabaster 07 Tinta de Toro  NaN   

        ReceivingDate  PurchasePrice  Quantity  Dollars  Classification  \
1109668    2016-07-04           6.93         7    48.51               1   
1112426    2016-07-09           6.93         6    41.58               1   
1116302    2016-07-07          91.83         1    91.83               2   

         OrdenCompraID  
1109668           2839  
1112426           2839  
1116302           2846

#### Busco en otras tablas los valores correctos de Size para Brand 3121, 5678 y 15365

In [235]:
# Filtrar la fila donde 'Brand' es igual a 3121
row_brand_3121 = df_PurchasePrices[df_PurchasePrices['Brand'] == 3121]

# Mostrar la fila
row_brand_3121


Brand               Description  Price   Size Volume  Classification  \
572   3121  Pinnacle Rainbow Sherbet   9.99  750mL    750               1   

     PurchasePrice  VendorNumber               VendorName  
572           6.93         12546  JIM BEAM BRANDS COMPANY

#### Verificamos si solo hay un size para cada Brand

In [237]:
# Agrupar por 'Brand' y contar los valores únicos de 'Size'
df_grouped = df_PurchasePrices.groupby('Brand')['Size'].nunique().reset_index()

# Filtrar los casos donde hay más de un valor único de 'Size'
df_multiple_size = df_grouped[df_grouped['Size'] > 1]

# Ver cuántos Brand tienen más de un Size
count_size= df_multiple_size.shape[0]

# Mostrar los Brand con más de un Size
print(df_multiple_size[['Brand', 'Size']])

# Mostrar cuántos Brand tienen más de un Size
print(f"Numero de Brand con más de un unico Size: {count_size}")


Empty DataFrame
Columns: [Brand, Size]
Index: []
Numero de Brand con más de un unico Size: 0


In [240]:
# Filtrar las filas donde la columna 'Brand' sea igual a '3121', '5678' o '15365'
brands_to_check = [3121, 5678, 15365]
df_filtered = df_PurchasePrices[df_PurchasePrices['Brand'].isin(brands_to_check)]

# Mostrar solo las columnas 'Brand' y 'Size' para las brands seleccionadas
df_filtered[['Brand', 'Size']].value_counts()


Brand  Size 
3121   750mL    1
5678   750mL    1
15365  750mL    1
Name: count, dtype: int64

In [241]:
# Llenar los espacios vacíos en la columna 'Size' con el valor '750mL'
Fact_Detalle_Compras.fillna({'Size':'750mL'}, inplace=True)

# Verificar que se hayan llenado los valores nulos
Fact_Detalle_Compras['Size'].info()


<class 'pandas.core.series.Series'>
RangeIndex: 2372474 entries, 0 to 2372473
Series name: Size
Non-Null Count    Dtype 
--------------    ----- 
2372474 non-null  object
dtypes: object(1)
memory usage: 18.1+ MB


In [242]:
total_nulos_detallecompras = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos_detallecompras)

Series([], dtype: int64)


### Guardamos nuestros dos archivos corregidos (llenado o retirando filas con vacios o nulos)

In [243]:
df_PurchasePrices.to_csv('PurchasePrices.csv', index=False)
Fact_Detalle_Compras.to_csv('Fact_DetalleCompras.csv', index=False)

#### Abrimos nuestros nuevos archivos

In [244]:
df_PurchasePrices= pd.read_csv('PurchasePrices.csv')
Fact_Detalle_Compras = pd.read_csv('Fact_DetalleCompras.csv')

### Inspeccionamos la tabla PurchasePrices para obtener la tabla Productos y vincularla con las tablas inventario y Compras y Ventas

In [245]:
df_PurchasePrices.shape

(12256, 9)

In [246]:
df_PurchasePrices.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorNumber', 'VendorName'],
      dtype='object')

In [247]:
df_PurchasePrices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12256 entries, 0 to 12255
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Brand           12256 non-null  int64  
 1   Description     12256 non-null  object 
 2   Price           12256 non-null  float64
 3   Size            12256 non-null  object 
 4   Volume          12256 non-null  float64
 5   Classification  12256 non-null  int64  
 6   PurchasePrice   12256 non-null  float64
 7   VendorNumber    12256 non-null  int64  
 8   VendorName      12256 non-null  object 
dtypes: float64(3), int64(3), object(3)
memory usage: 861.9+ KB


In [248]:
total_nulos = df_PurchasePrices.isnull().sum()[df_PurchasePrices.isnull().sum()>0]
total_nulos


Series([], dtype: int64)

In [249]:
df_PurchasePrices['Brand'].value_counts()

Brand
33052    1
30117    1
30039    1
28050    1
28029    1
        ..
75       1
72       1
63       1
62       1
58       1
Name: count, Length: 12256, dtype: int64

In [250]:
df_PurchasePrices['Classification'].value_counts()

Classification
2    8692
1    3564
Name: count, dtype: int64

In [251]:
df_PurchasePrices['Description'].value_counts()

Description
Jagermeister Liqueur            7
Bacardi Superior Rum            7
Jack Daniels No 7 Black         6
Capt Morgan Spiced Rum          6
Jim Beam                        6
                               ..
Priest Ranch Sauvignon Blanc    1
DLD Syrah -Les Gosses           1
Morgan Winery Albarino          1
Chateau D'Armailhac 2014        1
Clos Haut Peyraguey 2014        1
Name: count, Length: 11110, dtype: int64

In [252]:
df_PurchasePrices['Price'].value_counts()

Price
9.99      703
12.99     641
11.99     583
14.99     519
10.99     517
         ... 
472.49      1
221.99      1
293.99      1
409.99      1
667.99      1
Name: count, Length: 378, dtype: int64

In [253]:
df_PurchasePrices['PurchasePrice'].value_counts()

PurchasePrice
6.53      62
6.66      59
6.70      55
6.84      54
7.14      54
          ..
68.62      1
72.60      1
57.53      1
44.21      1
442.38     1
Name: count, Length: 2312, dtype: int64

In [255]:
df_PurchasePrices['Size'].value_counts()

Size
750mL          9823
1500mL          486
1750mL          433
1000mL          427
375mL           318
50mL            276
3000mL          104
187mL            52
500mL            51
187mL 4 Pk       48
5000mL           44
200mL            26
300mL            24
4000mL           16
720mL            11
1500ml           11
100mL             8
750mL 2 Pk        6
250mL 4 Pk        6
50mL 4 Pk         6
180mL             5
330mL             5
50mL 3 Pk         5
20000mL           5
200mL 3 Pk        4
6000mL            4
50mL 5 Pk         4
187mL 3 Pk        4
375mL 2 Pk        3
200mL 4 Pk        3
750mL 3 Pk        3
18000mL           3
250mL             3
150mL             2
100mL 4 Pk        2
375mL 3 Pk        2
750mL + 3/        2
750mL 6 Pk        2
600mL             2
560mL             2
20mL 5 Pk         1
750mL + 4/        1
1000mL 2 Pk       1
650mL             1
750mL 12 P        1
750mL + 1/        1
50mL 12 Pk        1
400mL             1
162.5mL           1
200mL 5 Pk     

In [256]:
df_PurchasePrices['VendorName'].value_counts()

VendorName
MARTIGNETTI COMPANIES          1637
ULTRA BEVERAGE COMPANY LLP      965
M S WALKER INC                  960
PERFECTA WINES                  897
E & J GALLO WINERY              527
                               ... 
TRUETT HURST                      1
STARK BREWING COMPANY             1
AAPER ALCOHOL & CHEMICAL CO       1
BLACK COVE BEVERAGES              1
LOYAL DOG WINERY                  1
Name: count, Length: 132, dtype: int64

In [257]:
df_PurchasePrices['VendorNumber'].value_counts()

VendorNumber
4425     1637
9165      965
9552      960
10754     897
3252      527
         ... 
99166       1
9099        1
54          1
90059       1
5083        1
Name: count, Length: 130, dtype: int64

In [258]:
df_PurchasePrices['Volume'].value_counts()

Volume
750.0      9840
1500.0      497
1750.0      433
1000.0      428
375.0       323
50.0        292
3000.0      104
187.0       104
500.0        51
5000.0       44
200.0        34
300.0        24
4000.0       16
720.0        11
100.0        10
250.0         9
20000.0       5
330.0         5
180.0         5
6000.0        4
18000.0       3
150.0         2
560.0         2
600.0         2
20.0          1
650.0         1
400.0         1
162.5         1
1100.0        1
19500.0       1
3750.0        1
9000.0        1
Name: count, dtype: int64

#### Se detecta que hay 130 VendorNumber pero 132 VendorName, se revisa y corrige en el dataframe df_PurchasePrices

In [259]:
# Agrupar por 'VendorNumber' y contar los valores únicos de 'VendorName'
df_grouped = df_PurchasePrices.groupby('VendorNumber')['VendorName'].nunique().reset_index()

# Filtrar los casos donde hay más de un valor único de 'VendorName'
df_multiple_vendorname = df_grouped[df_grouped['VendorName'] > 1]

# Ver cuántos VendorNumber tienen más de un VendorName
count_vendorname= df_multiple_vendorname.shape[0]

# Mostrar los VendorNumber con más de un VendorName
print(df_multiple_vendorname[['VendorNumber', 'VendorName']])

# Mostrar cuántos VendorNumber tienen más de un VendorName
print(f"Numero de VendorNumber con más de un unico VendorName: {count_vendorname}")


    VendorNumber  VendorName
20          1587           2
26          2000           2
Numero de VendorNumber con más de un unico VendorName: 2


In [261]:
# Filtrar la fila donde 'VendorNumber' es igual a 2000
row_VendorNumber_2000 = df_PurchasePrices[df_PurchasePrices['VendorNumber'] == 2000]

# Mostrar la fila
row_VendorNumber_2000['VendorName'].unique()

array(['SOUTHERN WINE & SPIRITS NE', 'SOUTHERN GLAZERS W&S OF NE'],
      dtype=object)

In [262]:
# Filtrar la fila donde 'VendorNumber' es igual a 1587
row_VendorNumber_1587 = df_PurchasePrices[df_PurchasePrices['VendorNumber'] == 1587]

# Mostrar la fila
row_VendorNumber_1587['VendorName'].unique()

array(['VINEYARD BRANDS INC', 'VINEYARD BRANDS LLC'], dtype=object)

#### Buscamos el nombre correcto del Vendor

In [260]:
# Filtrar las filas donde la columna 'VendorNumber' sea igual a '1587' y '2000'
names_to_check = [1587, 2000]
df_filtered = DimVendor[DimVendor['VendorNumber'].isin(names_to_check)]

# Mostrar solo las columnas 'VendorNumber' y 'VendorName' para las brands seleccionadas
df_filtered[['VendorNumber', 'VendorName']].value_counts()


VendorNumber  VendorName                
1587          VINEYARD BRANDS INC           1
2000          SOUTHERN WINE & SPIRITS NE    1
Name: count, dtype: int64

#### Corregimos el nombre en df_PurchasePrices

In [263]:
# Verificar si el nombre incorrecto está en la columna VendorName y corregirlo
df_PurchasePrices.loc[df_PurchasePrices['VendorName'] == 'SOUTHERN GLAZERS W&S OF NE', 'VendorName'] = 'SOUTHERN WINE & SPIRITS NE'


In [264]:
# Verificar si el nombre incorrecto está en la columna VendorName y corregirlo
df_PurchasePrices.loc[df_PurchasePrices['VendorName'] == 'VINEYARD BRANDS LLC', 'VendorName'] = 'VINEYARD BRANDS INC'


#### Verificamos la correccion

In [265]:
# Filtrar las filas donde la columna 'VendorNumber' sea igual a '1587' y '2000'
names_to_check = [1587, 2000]
df_filtered = df_PurchasePrices[df_PurchasePrices['VendorNumber'].isin(names_to_check)]

# Mostrar solo las columnas 'VendorNumber' y 'VendorName' para las brands seleccionadas
df_filtered[['VendorNumber', 'VendorName']].value_counts()


VendorNumber  VendorName                
2000          SOUTHERN WINE & SPIRITS NE    417
1587          VINEYARD BRANDS INC            87
Name: count, dtype: int64

### Seguimos con la revision de PurchasePrices

In [266]:
df_PurchasePrices.shape

(12256, 9)

In [267]:
total_nulos = df_PurchasePrices.isnull().sum()[df_PurchasePrices.isnull().sum()>0]
total_nulos


Series([], dtype: int64)

In [268]:
df_PurchasePrices['Brand'].value_counts()

Brand
33052    1
30117    1
30039    1
28050    1
28029    1
        ..
75       1
72       1
63       1
62       1
58       1
Name: count, Length: 12256, dtype: int64

In [269]:
df_PurchasePrices.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorNumber', 'VendorName'],
      dtype='object')

### Insertamos el VendorID en la tabla PurchasePrice, retiramos columnas redundantes, creamos el ID y cambiamos el nombre por tablaProducto

In [270]:
# Realizar un merge entre PurchasePrices con DimVendor basándose en la columna VendorNumber
Producto = df_PurchasePrices.merge(DimVendor[['VendorID','VendorNumber']], left_on='VendorNumber', right_on='VendorNumber', how='left')

### Eliminamos columnas redundantes

In [272]:
Producto.shape

(12256, 10)

In [271]:
Producto.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorNumber', 'VendorName', 'VendorID'],
      dtype='object')

In [274]:
# Elimino la columna 'VendorNumber' y 'VendorName'
Producto = Producto.drop(columns=['VendorNumber','VendorName'])


In [276]:
# Reviso si hay nulos
total_nulos = Producto.isnull().sum()[Producto.isnull().sum()>0]
print(total_nulos)


VendorID    5
dtype: int64


In [277]:
# Busco donde se encuentran y reviso las filas
Producto[Producto['VendorID'].isnull()]

Brand                  Description  Price    Size  Volume  \
382    2539  John Barr Red Scotch Whisky  19.99  1750mL  1750.0   
5017   3339                   Halo Vodka  19.99   750mL   750.0   
7164   4015     Ron Fortuna 8 Yr Old Rum  16.99   750mL   750.0   
7628   8263         Diep 9 Passion Fruit   9.99   750mL   750.0   
7638  15625     Mourvedre Bunnell Family  27.99   750mL   750.0   

      Classification  PurchasePrice  VendorID  
382                1          12.30       NaN  
5017               1          15.62       NaN  
7164               1          12.87       NaN  
7628               1           7.46       NaN  
7638               2          18.54       NaN

In [279]:
# Filtrar las filas donde la columna 'Brand' sea igual a '2539', '3339','4015', '8263' y '15625'
brands_to_check = [2539, 3339, 4015, 8263, 15625]
df_filtered = df_SalesFinal[df_SalesFinal['Brand'].isin(brands_to_check)]

# Mostrar solo las columnas 'Brand' y 'Size' para las brands seleccionadas
df_filtered[['Brand', 'VendorName','SalesDollars']].value_counts()


Brand  VendorName      SalesDollars
2539   WHYTE & MACKAY  31.98           1
4015   BERNIKO LLC     16.99           1
Name: count, dtype: int64

In [285]:
# Filtrar las filas donde la columna 'Brand' sea igual a '2539', '3339','4015', '8263' y '15625'
brands_to_check = [2539, 3339, 4015, 8263, 15625]
df_filtered = Fact_Detalle_Compras[Fact_Detalle_Compras['Brand'].isin(brands_to_check)]

# Mostrar solo las columnas 'Brand' y 'Size' para las brands seleccionadas
df_filtered[['Brand', 'Dollars']].value_counts()


Series([], Name: count, dtype: int64)

#### Eliminamos esas brand que no tienen vendors, pues no tienen historico de compras y las ventas no son significativas para analisis

In [286]:
# Eliminar las filas con los valores 2539, 3339, 4015, 8263, y 15625 en la columna 'Brand'
Producto = Producto[~Producto['Brand'].isin([2539, 3339, 4015, 8263, 15625])]


In [287]:
# Reviso si hay nulos
total_nulos = Producto.isnull().sum()[Producto.isnull().sum()>0]
print(total_nulos)

Series([], dtype: int64)


In [288]:
# Reviso si hay nulos
total_nulos = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos)


Series([], dtype: int64)


In [289]:
# Reviso si hay nulos
total_nulos = Ordenes_Compra.isnull().sum()[Ordenes_Compra.isnull().sum()>0]
print(total_nulos)


Series([], dtype: int64)


In [290]:
Producto.shape

(12251, 8)

In [292]:
Producto.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12251 entries, 0 to 12255
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Brand           12251 non-null  int64  
 1   Description     12251 non-null  object 
 2   Price           12251 non-null  float64
 3   Size            12251 non-null  object 
 4   Volume          12251 non-null  float64
 5   Classification  12251 non-null  int64  
 6   PurchasePrice   12251 non-null  float64
 7   VendorID        12251 non-null  float64
dtypes: float64(4), int64(2), object(2)
memory usage: 861.4+ KB


#### Creamos el ProductoID en la nueva dataframe Producto

In [293]:
# Generar ProductoID en Producto
Producto['ProductoID'] = Producto.index + 1

In [294]:
Producto['Brand'].value_counts() # Es la PK

Brand
46964    1
58       1
62       1
63       1
72       1
        ..
211      1
192      1
178      1
175      1
173      1
Name: count, Length: 12251, dtype: int64

In [295]:
Producto.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorID', 'ProductoID'],
      dtype='object')

### Insertamos el ProductoID en la tabla DetalleCompras, retiramos columnas redundantes, y revisamos nulos

In [296]:
# Realizar un merge entre Fact_DetalleCompras con Producto basándose en la columna Brand
Fact_Detalle_Compras = Fact_Detalle_Compras.merge(Producto[['Brand','ProductoID']], left_on='Brand', right_on='Brand', how='left')

### Eliminamos columnas redundantes

In [297]:
Fact_Detalle_Compras.shape

(2372474, 12)

In [298]:
Fact_Detalle_Compras.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'ReceivingDate',
       'PurchasePrice', 'Quantity', 'Dollars', 'Classification',
       'OrdenCompraID', 'ProductoID'],
      dtype='object')

In [299]:
# Elimino las columnas redundantes
Fact_Detalle_Compras = Fact_Detalle_Compras.drop(columns=['Brand', 'Description', 'Size', 'Classification'])


In [301]:
# Reviso si hay nulos
total_nulos = Fact_Detalle_Compras.isnull().sum()[Fact_Detalle_Compras.isnull().sum()>0]
print(total_nulos)


Series([], dtype: int64)


In [302]:
Fact_Detalle_Compras.columns

Index(['InventoryId', 'Store', 'ReceivingDate', 'PurchasePrice', 'Quantity',
       'Dollars', 'OrdenCompraID', 'ProductoID'],
      dtype='object')

In [303]:
Fact_Detalle_Compras.shape

(2372474, 8)

### Guardamos nuestros dos archivos modificados

In [304]:
Producto.to_csv('Dim_Producto.csv', index=False)
Fact_Detalle_Compras.to_csv('Fact_DetalleCompras.csv', index=False)

### Volvemos a abrir nuestros archivos para seguir trabajando

In [305]:
Producto= pd.read_csv('Dim_Producto.csv')
Fact_Detalle_Compras =pd.read_csv('Fact_DetalleCompras.csv')

### Revisamos las tablas Inventario

In [306]:
df_BegInv.columns

Index(['InventoryId', 'Store', 'City', 'Brand', 'Description', 'Size',
       'onHand', 'Price', 'startDate'],
      dtype='object')

In [307]:
df_BegInv.shape

(206529, 9)

In [319]:
df_EndInv.shape

(224489, 9)

In [326]:
df_BegInv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206529 entries, 0 to 206528
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InventoryId  206529 non-null  object 
 1   Store        206529 non-null  int64  
 2   City         206529 non-null  object 
 3   Brand        206529 non-null  int64  
 4   Description  206529 non-null  object 
 5   Size         206529 non-null  object 
 6   onHand       206529 non-null  int64  
 7   Price        206529 non-null  float64
 8   startDate    206529 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 14.2+ MB


In [327]:
df_EndInv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 224489 entries, 0 to 224488
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InventoryId  224489 non-null  object 
 1   Store        224489 non-null  int64  
 2   City         223205 non-null  object 
 3   Brand        224489 non-null  int64  
 4   Description  224489 non-null  object 
 5   Size         224489 non-null  object 
 6   onHand       224489 non-null  int64  
 7   Price        224489 non-null  float64
 8   endDate      224489 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 15.4+ MB


In [313]:
df_BegInv['InventoryId'].value_counts() # esta es la PK

InventoryId
79_BALLYMENA_90089    1
79_BALLYMENA_46568    1
79_BALLYMENA_46551    1
79_BALLYMENA_46489    1
79_BALLYMENA_46476    1
                     ..
1_HARDERSFIELD_75     1
1_HARDERSFIELD_72     1
1_HARDERSFIELD_63     1
1_HARDERSFIELD_62     1
1_HARDERSFIELD_60     1
Name: count, Length: 206529, dtype: int64

In [320]:
df_EndInv['InventoryId'].value_counts() # esta es la PK

InventoryId
81_PEMBROKE_90604     1
1_HARDERSFIELD_58     1
1_HARDERSFIELD_62     1
1_HARDERSFIELD_63     1
1_HARDERSFIELD_72     1
                     ..
1_HARDERSFIELD_192    1
1_HARDERSFIELD_178    1
1_HARDERSFIELD_175    1
1_HARDERSFIELD_171    1
1_HARDERSFIELD_159    1
Name: count, Length: 224489, dtype: int64

In [308]:
df_BegInv['Brand'].value_counts()

Brand
77       79
8678     79
8691     79
3344     79
3352     79
         ..
14193     1
24229     1
24161     1
23288     1
23135     1
Name: count, Length: 8094, dtype: int64

In [321]:
df_EndInv['Brand'].value_counts()

Brand
3654     80
3488     80
2122     80
3724     80
3502     80
         ..
4114      1
3648      1
1605      1
24109     1
17444     1
Name: count, Length: 9653, dtype: int64

In [309]:
df_BegInv['City'].value_counts()

City
MOUNTMEND       14264
EANVERNESS      13011
DONCASTER       10356
HORNSEY          9997
GOULCREST        7930
                ...  
AYLESBURY        1370
KNIFE'S EDGE     1343
VERITAS          1323
LUNDY            1291
EASTHAVEN        1225
Name: count, Length: 67, dtype: int64

In [322]:
df_EndInv['City'].value_counts()

City
MOUNTMEND       18065
EANVERNESS      12879
DONCASTER       10565
HORNSEY         10202
GOULCREST        8061
                ...  
LUTON            1425
KNIFE'S EDGE     1398
AYLESBURY        1366
VERITAS          1316
LUNDY            1302
Name: count, Length: 67, dtype: int64

In [311]:
df_BegInv['Store'].value_counts()

Store
34    6026
69    5400
73    5241
67    5179
76    5115
      ... 
45    1323
16    1291
78    1225
28    1046
3      481
Name: count, Length: 79, dtype: int64

In [325]:
df_EndInv['Store'].value_counts()

Store
50    7111
34    6442
69    5724
73    5445
74    5222
      ... 
45    1316
16    1302
46    1284
28    1049
3      463
Name: count, Length: 80, dtype: int64

In [318]:
df_BegInv['startDate'].value_counts()

startDate
2016-01-01    206529
Name: count, dtype: int64

In [328]:
df_EndInv['endDate'].value_counts()

endDate
2016-12-31    224489
Name: count, dtype: int64

### Revisamos primero los casos de la columna city con datos incompletos

In [329]:
# Reviso si hay nulos
total_nulos = df_EndInv.isnull().sum()[df_EndInv.isnull().sum()>0]
print(total_nulos)


City    1284
dtype: int64


In [330]:
df_EndInv[df_EndInv['City'].isnull()]

InventoryId  Store City  Brand                   Description   Size  \
113895      46__58     46  NaN     58   Gekkeikan Black & Gold Sake  750mL   
113896      46__62     46  NaN     62      Herradura Silver Tequila  750mL   
113897      46__63     46  NaN     63    Herradura Reposado Tequila  750mL   
113898      46__77     46  NaN     77   Three Olives Espresso Vodka  750mL   
113899     46__106     46  NaN    106      Mr Boston Peach Schnapps  Liter   
...            ...    ...  ...    ...                           ...    ...   
115174   46__46447     46  NaN  46447         Gascon Malbec Mendoza  750mL   
115175   46__46458     46  NaN  46458      Layer Cake Barosa Shiraz  750mL   
115176   46__46476     46  NaN  46476          Tilia Malbec Mendoza  750mL   
115177   46__46764     46  NaN  46764  Clayhouse Adobe Red Paso Rbl  750mL   
115178   46__46830     46  NaN  46830         Pacific Rim Sweet Rsl  750mL   

        onHand  Price     endDate  
113895       0  12.99  2016-12-31  
113896       0  36.99  2016-12-31  
113897       0  38.99  2016-12-31  
113898       0  14.99  2016-12-31  
113899       0   4.49  2016-12-31  
...        ...    ...         ...  
115174       0  10.99  2016-12-31  
115175       0  15.99  2016-12-31  
115176       0   9.99  2016-12-31  
115177       0  11.99  2016-12-31  
115178       0   8.99  2016-12-31  

[1284 rows x 9 columns]

In [332]:
# Verificamos si la ciudad vacia es solo para la Store 46
# Filtrar filas donde la columna 'City' está vacía (NaN)
empty_city_rows = df_EndInv[df_EndInv['City'].isna()]

# Verificar si todas las filas con 'City' vacía corresponden a 'Store' 46
all_store_46 = empty_city_rows['Store'].eq(46).all()

# Mostrar el resultado
if all_store_46:
    print("Todos los valores vacíos en 'City' corresponden a la Store 46.")
else:
    print("No todos los valores vacíos en 'City' corresponden a la Store 46.")


Todos los valores vacíos en 'City' corresponden a la Store 46.


### Buscamos un valor de City para Store 46 y rellenamos los vacios

In [337]:
# Filtrar las filas donde 'Store' es 46 y 'City' no es NaN (valor válido)
valid_city_for_store_46 = df_BegInv.loc[(df_BegInv['Store'] == 46) & (df_BegInv['City'].notna()), 'City'].unique()


In [338]:
print(valid_city_for_store_46)

['TYWARDREATH']


In [339]:
# Rellenar los valores vacíos en 'City' para 'Store' 46 con el valor correcto
df_EndInv.loc[(df_EndInv['Store'] == 46) & (df_EndInv['City'].isna()), 'City'] = 'TYWARDREATH'

print(f"Los valores vacíos de 'City' para Store 46 se han rellenado con: {'TYWARDREATH'}")

Los valores vacíos de 'City' para Store 46 se han rellenado con: TYWARDREATH


In [340]:
# Reviso si hay nulos
total_nulos = df_EndInv.isnull().sum()[df_EndInv.isnull().sum()>0]
print(total_nulos)


Series([], dtype: int64)


In [341]:
# Filtrar nuevamente la fila donde 'Store' es igual a 46
row_Store_46 = df_EndInv[df_EndInv['Store'] == 46]

# Mostrar la fila
row_Store_46['City'].unique()

array(['TYWARDREATH'], dtype=object)

In [343]:
# Revisar el InventoryId que no queda unido correctamente con la City 'Tywardreath'
# Filtrar nuevamente la fila donde 'Store' es igual a 46
row_Store_46 = df_EndInv[df_EndInv['Store'] == 46]

# Mostrar la fila
row_Store_46['InventoryId'].unique()

array(['46__58', '46__62', '46__63', ..., '46__46476', '46__46764',
       '46__46830'], dtype=object)

### Guardamos nuestros avances con el dataframe df_EndInv

In [344]:
df_EndInv.to_csv('Dim_Inventario_end_year.csv', index=False)

### Abrimos nuevamente

In [345]:
df_EndInv= pd.read_csv('Dim_Inventario_end_year.csv')

### Revisamos la tabla Producto para insertala en Detalle Ventas e Inventario

In [368]:
Producto.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorID', 'ProductoID'],
      dtype='object')

In [374]:
df_EndInv.columns

Index(['InventoryId', 'Store', 'City', 'Brand', 'Description', 'Size',
       'onHand', 'Price', 'endDate'],
      dtype='object')

In [375]:
df_BegInv.columns

Index(['InventoryId', 'Store', 'City', 'Brand', 'Description', 'Size',
       'onHand', 'Price', 'startDate'],
      dtype='object')

In [379]:
total_nulos_inv = df_EndInv.isnull().sum()[df_EndInv.isnull().sum()>0]
print(total_nulos_inv)

Series([], dtype: int64)


In [380]:
total_nulos_inv = df_BegInv.isnull().sum()[df_BegInv.isnull().sum()>0]
print(total_nulos_inv)

Series([], dtype: int64)


In [376]:
df_SalesFinal.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNo', 'VendorName'],
      dtype='object')

In [377]:
df_SalesFinal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 14 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   InventoryId     1048575 non-null  object 
 1   Store           1048575 non-null  int64  
 2   Brand           1048575 non-null  int64  
 3   Description     1048575 non-null  object 
 4   Size            1048575 non-null  object 
 5   SalesQuantity   1048575 non-null  int64  
 6   SalesDollars    1048575 non-null  float64
 7   SalesPrice      1048575 non-null  float64
 8   SalesDate       1048575 non-null  object 
 9   Volume          1048575 non-null  int64  
 10  Classification  1048575 non-null  int64  
 11  ExciseTax       1048575 non-null  float64
 12  VendorNo        1048575 non-null  int64  
 13  VendorName      1048575 non-null  object 
dtypes: float64(3), int64(6), object(5)
memory usage: 112.0+ MB


In [378]:
total_nulos_salesfinal = df_SalesFinal.isnull().sum()[df_SalesFinal.isnull().sum()>0]
print(total_nulos_salesfinal)

Series([], dtype: int64)


#### Inserto ProductoID en BegInv, EndInv, SalesFinal (y se les cambia el nombre a las 3 tablas)

In [382]:
# Realizar un merge entre Producto y BegInv, EndInv y SalesFinal basándose en Brand
Fact_Detalle_Ventas = df_SalesFinal.merge(Producto[['ProductoID','Brand']], left_on='Brand', right_on='Brand', how='left')
DimInv_EndYear = df_EndInv.merge(Producto[['ProductoID','Brand']], left_on='Brand', right_on='Brand', how='left')
DimInv_BegYear = df_BegInv.merge(Producto[['ProductoID','Brand']], left_on='Brand', right_on='Brand', how='left')


#### Eliminamos las columnas redundantes

In [383]:
# Elimino las columnas redundantes
Fact_Detalle_Ventas = Fact_Detalle_Ventas.drop(columns=['Brand', 'Description', 'Size', 'Volume', 'Classification','VendorNo', 'VendorName'])
DimInv_EndYear = DimInv_EndYear.drop(columns=['Brand', 'Description', 'Size', 'Price'])
DimInv_BegYear = DimInv_BegYear.drop(columns=['Brand', 'Description', 'Size', 'Price'])



#### Reviso que no tengan nulos, En caso de tener nulos, los eliminamos

In [384]:

total_nulos_ventas = Fact_Detalle_Ventas.isnull().sum()[Fact_Detalle_Ventas.isnull().sum()>0]
print(total_nulos_ventas)

ProductoID    2
dtype: int64


In [386]:
total_nulos_endinv = DimInv_EndYear.isnull().sum()[DimInv_EndYear.isnull().sum()>0]
print(total_nulos_endinv)


Series([], dtype: int64)


In [387]:
total_nulos_beginv = DimInv_BegYear.isnull().sum()[DimInv_BegYear.isnull().sum()>0]
print(total_nulos_beginv)


ProductoID    3
dtype: int64


#### Revisamos los que tienen nulos y los eliminamos de ser necesario

In [391]:
# Filtrar las filas donde 'ProductoID' es nulo (NaN)
filas_nulas_productoID = Fact_Detalle_Ventas[Fact_Detalle_Ventas['ProductoID'].isnull()]

# Mostrar las filas
filas_nulas_productoID


InventoryId  Store  SalesQuantity  SalesDollars  SalesPrice  \
374815   41_LARNWICK_2539     41              2         31.98       15.99   
873327  76_DONCASTER_4015     76              1         16.99       16.99   

        SalesDate  ExciseTax  ProductoID  
374815  1/11/2016       3.67         NaN  
873327  1/29/2016       0.79         NaN

In [392]:
# Filtrar las filas donde 'ProductoID' es nulo (NaN)
filas_nulas_productoID2 = DimInv_BegYear[DimInv_BegYear['ProductoID'].isnull()]

# Mostrar las filas
filas_nulas_productoID2


InventoryId  Store       City  onHand   startDate  ProductoID
81186   34_PITMERDEN_15625     34  PITMERDEN       2  2016-01-01         NaN
98426     41_LARNWICK_2539     41   LARNWICK       2  2016-01-01         NaN
195641   76_DONCASTER_4015     76  DONCASTER       1  2016-01-01         NaN

#### Elimino los Nulos

In [393]:
# Eliminar las filas donde 'ProductoID' es nulo
Fact_Detalle_Ventas = Fact_Detalle_Ventas.dropna(subset=['ProductoID'])
DimInv_BegYear = DimInv_BegYear.dropna(subset=['ProductoID'])


In [394]:
total_nulos_ventas = Fact_Detalle_Ventas.isnull().sum()[Fact_Detalle_Ventas.isnull().sum()>0]
print(total_nulos_ventas)

Series([], dtype: int64)


In [395]:
total_nulos_beginv = DimInv_BegYear.isnull().sum()[DimInv_BegYear.isnull().sum()>0]
print(total_nulos_beginv)


Series([], dtype: int64)


In [396]:
Fact_Detalle_Ventas.columns

Index(['InventoryId', 'Store', 'SalesQuantity', 'SalesDollars', 'SalesPrice',
       'SalesDate', 'ExciseTax', 'ProductoID'],
      dtype='object')

In [397]:
DimInv_BegYear.columns

Index(['InventoryId', 'Store', 'City', 'onHand', 'startDate', 'ProductoID'], dtype='object')

In [398]:
DimInv_EndYear.columns

Index(['InventoryId', 'Store', 'City', 'onHand', 'endDate', 'ProductoID'], dtype='object')

In [399]:
Producto.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorID', 'ProductoID'],
      dtype='object')

## Revisamos los tipos de datos de nuestras tablas

In [408]:
DimInv_BegYear.dtypes

InventoryId     object
Store            int64
City            object
onHand           int64
startDate       object
ProductoID     float64
dtype: object

In [402]:
DimInv_EndYear.dtypes

InventoryId    object
Store           int64
City           object
onHand          int64
endDate        object
ProductoID      int64
dtype: object

In [403]:
DimVendor.dtypes

VendorNumber     int64
VendorName      object
VendorID         int64
dtype: object

In [404]:
Ordenes_Compra.dtypes

InvoiceDate       object
PONumber           int64
PODate            object
PayDate           object
Quantity           int64
Dollars          float64
Freight          float64
VendorID           int64
OrdenCompraID      int64
dtype: object

In [405]:
Producto.dtypes

Brand               int64
Description        object
Price             float64
Size               object
Volume            float64
Classification      int64
PurchasePrice     float64
VendorID          float64
ProductoID          int64
dtype: object

In [406]:
Fact_Detalle_Ventas.dtypes

InventoryId       object
Store              int64
SalesQuantity      int64
SalesDollars     float64
SalesPrice       float64
SalesDate         object
ExciseTax        float64
ProductoID       float64
dtype: object

In [407]:
Fact_Detalle_Compras.dtypes

InventoryId       object
Store              int64
ReceivingDate     object
PurchasePrice    float64
Quantity           int64
Dollars          float64
OrdenCompraID      int64
ProductoID         int64
dtype: object

### Cambiar a tipo fecha las columnas de mis dataframes

In [ ]:
# Convertir todas las columnas que contienen 'Date' en su nombre a formato de fecha
Fact_Detalle_Ventas[Fact_Detalle_Ventas.columns[Fact_Detalle_Ventas.columns.str.contains('Date')]] = Fact_Detalle_Ventas[Fact_Detalle_Ventas.columns[Fact_Detalle_Ventas.columns.str.contains('Date')]].apply(pd.to_datetime)
Fact_Detalle_Compras[Fact_Detalle_Compras.columns[Fact_Detalle_Compras.columns.str.contains('Date')]] = Fact_Detalle_Compras[Fact_Detalle_Compras.columns[Fact_Detalle_Compras.columns.str.contains('Date')]].apply(pd.to_datetime)
Ordenes_Compra[Ordenes_Compra.columns[Ordenes_Compra.columns.str.contains('Date')]] = Ordenes_Compra[Ordenes_Compra.columns[Ordenes_Compra.columns.str.contains('Date')]].apply(pd.to_datetime)
DimInv_BegYear[DimInv_BegYear.columns[DimInv_BegYear.columns.str.contains('Date')]] = DimInv_BegYear[DimInv_BegYear.columns[DimInv_BegYear.columns.str.contains('Date')]].apply(pd.to_datetime)
DimInv_EndYear[DimInv_EndYear.columns[DimInv_EndYear.columns.str.contains('Date')]] = DimInv_EndYear[DimInv_EndYear.columns[DimInv_EndYear.columns.str.contains('Date')]].apply(pd.to_datetime)



#### Eliminamos una columna redundante adicional en Detalle_Compras

In [411]:
# Elimino las columnas redundantes
Fact_Detalle_Compras = Fact_Detalle_Compras.drop(columns=['PurchasePrice' ])

### Cambiar a tipo entero algunas columnas

In [ ]:
# Redondear los valores de las columnas y luego convertir a int64
Fact_Detalle_Ventas['ProductoID'] = Fact_Detalle_Ventas['ProductoID'].round().astype('int64')
DimInv_BegYear['ProductoID'] = DimInv_BegYear['ProductoID'].round().astype('int64')


### Revisamos nuevamente los tipos de tados de nuestras tablas

In [422]:
DimInv_BegYear.dtypes

InventoryId            object
Store                   int64
City                   object
onHand                  int64
startDate      datetime64[ns]
ProductoID              int64
dtype: object

In [423]:
DimInv_EndYear.dtypes

InventoryId            object
Store                   int64
City                   object
onHand                  int64
endDate        datetime64[ns]
ProductoID              int64
dtype: object

In [424]:
DimVendor.dtypes

VendorNumber     int64
VendorName      object
VendorID         int64
dtype: object

In [425]:
Ordenes_Compra.dtypes

InvoiceDate      datetime64[ns]
PONumber                  int64
PODate           datetime64[ns]
PayDate          datetime64[ns]
Quantity                  int64
Dollars                 float64
Freight                 float64
VendorID                  int64
OrdenCompraID             int64
dtype: object

In [426]:
Producto.dtypes

Brand               int64
Description        object
Price             float64
Size               object
Volume            float64
Classification      int64
PurchasePrice     float64
VendorID          float64
ProductoID          int64
dtype: object

In [427]:
Fact_Detalle_Ventas.dtypes

InventoryId              object
Store                     int64
SalesQuantity             int64
SalesDollars            float64
SalesPrice              float64
SalesDate        datetime64[ns]
ExciseTax               float64
ProductoID                int64
dtype: object

In [428]:
Fact_Detalle_Compras.dtypes

InventoryId              object
Store                     int64
ReceivingDate    datetime64[ns]
Quantity                  int64
Dollars                 float64
OrdenCompraID             int64
ProductoID                int64
dtype: object

## Guardar todas nuestras tablas

In [429]:
DimVendor.to_csv('Dim_Vendor.csv',index=False)
DimInv_BegYear.to_csv('Fact_InvInicioYear.csv',index=False)
DimInv_EndYear.to_csv('Fact_InvFinYear.csv',index=False)
Ordenes_Compra.to_csv('Dim_OC.csv',index=False)
Producto.to_csv('Dim_Producto.csv',index=False)
Fact_Detalle_Ventas.to_csv('Fact_Ventas.csv',index=False)
Fact_Detalle_Compras.to_csv('Fact_Compras.csv',index=False)



### Modificacion a la tabla Orden_Compra

In [2]:
Ordenes_Compra=pd.read_csv('Dim_OC.csv')

In [4]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'VendorID', 'OrdenCompraID'],
      dtype='object')

In [3]:
DimVendor=pd.read_csv('Dim_Vendor.csv')

In [5]:
DimVendor.columns

Index(['VendorNumber', 'VendorName', 'VendorID'], dtype='object')

In [6]:
# Realizar un merge entre df_PurchasesFinal y df_SalesFinal y Detalle_brand basándose en BrandID y Brand
Ordenes_Compra = Ordenes_Compra.merge(DimVendor[['VendorNumber', 'VendorName', 'VendorID']], left_on='VendorID', right_on='VendorID', how='left')


In [7]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'VendorID', 'OrdenCompraID', 'VendorNumber', 'VendorName'],
      dtype='object')

#### Borramos columna VendorID

In [8]:
Ordenes_Compra = Ordenes_Compra.drop(columns=['VendorID'])

In [9]:
Ordenes_Compra.columns

Index(['InvoiceDate', 'PONumber', 'PODate', 'PayDate', 'Quantity', 'Dollars',
       'Freight', 'OrdenCompraID', 'VendorNumber', 'VendorName'],
      dtype='object')

#### Verificamos si hay nulos

In [10]:
total_nulos_oc = Ordenes_Compra.isnull().sum()[Ordenes_Compra.isnull().sum()>0]
print(total_nulos_oc)


Series([], dtype: int64)


#### Verificamos tipos de columnas

In [11]:
Ordenes_Compra.dtypes

InvoiceDate       object
PONumber           int64
PODate            object
PayDate           object
Quantity           int64
Dollars          float64
Freight          float64
OrdenCompraID      int64
VendorNumber       int64
VendorName        object
dtype: object

#### Cambiamos tipos de Datos a Fecha

In [12]:
# Convertir todas las columnas que contienen 'Date' en su nombre a formato de fecha
Ordenes_Compra[Ordenes_Compra.columns[Ordenes_Compra.columns.str.contains('Date')]] = Ordenes_Compra[Ordenes_Compra.columns[Ordenes_Compra.columns.str.contains('Date')]].apply(pd.to_datetime)


#### Verificamos nuevamente tipo de datos

In [13]:
Ordenes_Compra.dtypes

InvoiceDate      datetime64[ns]
PONumber                  int64
PODate           datetime64[ns]
PayDate          datetime64[ns]
Quantity                  int64
Dollars                 float64
Freight                 float64
OrdenCompraID             int64
VendorNumber              int64
VendorName               object
dtype: object

### Guardamos nuestra tabla Ordenes_Compra

In [14]:
Ordenes_Compra.to_csv('Dim_OC.csv',index=False)

### Modificacion del tipo de dato de la tabla Producto

In [1]:
import pandas as pd

In [2]:
Producto=pd.read_csv('Dim_Producto.csv')

In [3]:
Producto.dtypes

Brand               int64
Description        object
Price             float64
Size               object
Volume            float64
Classification      int64
PurchasePrice     float64
VendorID          float64
ProductoID          int64
dtype: object

### Cambiar a tipo entero VendorID

In [4]:
# Redondear los valores de las columnas y luego convertir a int64
Producto['VendorID'] = Producto['VendorID'].round().astype('int64')


In [5]:
# Revisamos nuevamente
Producto.dtypes

Brand               int64
Description        object
Price             float64
Size               object
Volume            float64
Classification      int64
PurchasePrice     float64
VendorID            int64
ProductoID          int64
dtype: object

## Guardamos nuestra tabla Producto

In [6]:
Producto.to_csv('Dim_Producto.csv',index=False)